runtime/src/iree/builtins/ukernel/arch/arm_64/mmt4d_arm_64_base.c

```diff
@@ -264,3 +264,120 @@ IREE_UK_MMT4D_TILE_FUNC_IMPL_FOR_M0_1_2_4_8(
     iree_uk_mmt4d_tile_s8s8s32_2x8x1_arm_64,
     iree_uk_mmt4d_tile_s8s8s32_4x8x1_arm_64,
     iree_uk_mmt4d_tile_s8s8s32_8x8x1_arm_64)
+
+static inline void iree_uk_mmt4d_tile_s8s4s32_1x16x2_to_4x16x2_arm_64(
+    void* IREE_UK_RESTRICT out_tile, const void* IREE_UK_RESTRICT lhs_panel,
+    const void* IREE_UK_RESTRICT rhs_panel,
+    const iree_uk_mmt4d_params_t* params, int M0) {
+  IREE_UK_ASSERT(M0 >= 1 && M0 <= 4 && iree_uk_is_po2_u32(M0));
+  IREE_UK_ASSERT(!(params->K0 % 2));
+  const iree_uk_int8_t* IREE_UK_RESTRICT lhs_ptr = lhs_panel;
+  const iree_uk_int8_t* IREE_UK_RESTRICT rhs_ptr = rhs_panel;
+
+  iree_uk_int32_t* IREE_UK_RESTRICT out_ptr = out_tile;
+  int32x4_t acc[16];
+  if (params->flags & IREE_UK_FLAG_MMT4D_ACCUMULATE) {
+    for (int i = 0; i < 4 * M0; ++i) {
+      acc[i] = vld1q_s32(out_ptr + 4 * i);
+    }
+  } else {
+    for (int i = 0; i < 4 * M0; ++i) {
+      acc[i] = vdupq_n_s32(0);
+    }
+  }
+  const int8x8_t vmask0f = vmov_n_s8(INT8_C(0x0F));
+  for (int k = 0; k < params->K; ++k) {
+    // Handle 1x16x2.
+    int16x8_t lhs = vmovl_s8(vld1_s8(lhs_ptr));
+    lhs_ptr += 2 * M0;
+
+    int8x8_t rhs_0 = vld1_s8(rhs_ptr);
+    rhs_ptr += 8;
+    int8x8_t rhs_1 = vld1_s8(rhs_ptr);
+    rhs_ptr += 8;
+
+    int16x8_t rhs_a = vmovl_s8(vand_s8(rhs_0, vmask0f));
```

Comment:
Missing context but where is the sign extension happening here?

Reply:
I would expect a shift left by 4 and then an arithmetic shift right by 4 or something like that

Reply:
Actually this is a great catch @dcaballe. I have now read carefully the code and something doesn't add up:
* This kernel is for the case where the RHS is *signed* 4-bit, as indicated by the letter `s` in `s4`.
* So the top bit in the 4-bit values is a sign and extension to 8-bit needs to be sign-extension.
* But the `vand_s8(rhs_0, vmask0f)` is performing zero-extension, not sign-extension to 8-bit. Then the `vmovl_s8` performs a sign-extention to 16-bit, but because it was fed the zero-extended 8-bit value, it never sees a sign bit and amounts to another zero-extension.
* I don't see any arithmetic elsewhere in this kernel that would amount to fixing this up.
* Therefore, to my best understanding, this is a `s8u4s32` kernel, not a `s8s4s32` kernel.

Reply:
Good catch, I dug deeper here to understand what was being done in XNNPack. I fixed the reference code to be `s8s4s32` and also added some comments in the ARM ukernel documenting how int4's are handled.